File: Level 3/가장 먼 노드.swift
func solution(_ n:Int, _ edge:[[Int]]) -> Int {
    var graph : [[Int]] = Array(repeating: [], count: n + 1)
    
    // graph에 연결된 노드 정보 담기
    // [[], [3, 2], [3, 1, 4, 5], [6, 4, 2, 1], [3, 2], [2], [3]]
    for e in edge {
        graph[e[0]].append(e[1])
        graph[e[1]].append(e[0])
    }
    
    return dfs(graph: graph)
}

func dfs(graph: [[Int]]) -> Int {
    var visited = Array(repeating: 0, count: graph.count) // 방문한 노드의 깊이를 저장할 배열
    var queue : [Int] = [] // 현재 탐색중인 노드 번호를 담을 큐
    queue.append(1) // 1번 노드부터 탐색
    
    while !queue.isEmpty {
        let now = queue.removeFirst() // 먼저 들어온 노드부터 제거 (선입선출)
        for next in graph[now] {
            if visited[next] == 0 { // 방문하지 않았던 노드인 경우
                visited[next] = visited[now] + 1 // 방문 처리
                queue.append(next) // 연결된 노드 큐에 담기
            }
        }
    }
    visited[1] = 0 // 1번 노드는 초기화
    
    let max = visited.max() // 최대 노드 깊이
    return visited.filter{ max == $0 }.count // result에서 max와 같은 깊이 count
}

File: Level 3/가장 먼 노드.java
import java.util.*;

class Solution {
    static ArrayList<ArrayList<Integer>> graph = new ArrayList<>();
    static boolean[] visited;
    
    public int solution(int n, int[][] edge) {
        visited = new boolean[n + 1];
        int answer = 0;

        for (int i = 0; i <= n; i++) {
            graph.add(i, new ArrayList<>());
        }

        for (int i = 0; i < edge.length; i++) { // 양방향 추가
            graph.get(edge[i][0]).add(edge[i][1]);
            graph.get(edge[i][1]).add(edge[i][0]);
        }
        
        answer = bfs();
        return answer;
    }
    
    public static int bfs() {
        Queue<Integer> queue = new LinkedList<>(); // 방문한 노드의 깊이를 저장할 배열
        queue.add(1); // 1번 노드부터 탐색
        visited[1] = true; // 1번 노드 방문 처리
        
        int cnt = 0;
        while (true) {
            Queue<Integer> temp = new LinkedList<>();

            while (!queue.isEmpty()) {
                int now = queue.poll();
                for (int adj : graph.get(now)) {
                    if (!visited[adj]) {
                        temp.add(adj);
                        visited[adj] = true;
                    }
                }
            }

            if (temp.isEmpty()) {
                break;
            }
            queue.addAll(temp);
            cnt = temp.size(); // 가장 먼 노드의 개수
        }

        return cnt;
    }
}
